Implement write methods in VaultFormat8ProviderDecorator

SailReal opened the pull request.

overheadhunter committed:
reformatted code (tabs instead of spaces) and fixed javadoc
[ci skip]

SailReal committed:
Implement write operations in VaultFormat8ProviderDecorator

SailReal committed:
Use new EncryptingReadableByteChannel in write-method

This avoids first writing the cipthertext into a temp file

SailReal committed:
Merge branch 'develop' into feature/crypto-decorator

SailReal committed:
Evict folders and sub folders from DirIdCache while deleting

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/vaultformat8/DirectoryIdCache.java:
```diff
@@ -32,8 +33,19 @@
+	}
+
+	public CompletionStage<Void> evictFolderIncludingChilds(CloudPath cleartextPath) {
+		return CompletableFuture.supplyAsync(() -> {
+			var subPaths = cache.keySet().stream().filter(path -> path.startsWith(cleartextPath)).collect(Collectors.toSet());
+			subPaths.forEach(cache::remove);
+			return null;
+		});
```
I don't think this qualifies for async execution. Spawning a background task is probably more expensive than iterating over a few thousand entries (worst case!).

Good old Iterator-over-EntrySet-loop should be fine and is easier to debug, too (no race conditions).

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/vaultformat8/DirectoryIdCache.java:
```diff
@@ -32,8 +33,19 @@
+		cache.remove(cleartextPath);
+	}
+
+	public CompletionStage<Void> evictFolderIncludingChilds(CloudPath cleartextPath) {
```
Correct: `evictIncludingChildren`. Even more correct: `evictIncludingDescendants` (all generations^^).

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/vaultformat8/VaultFormat8ProviderDecorator.java:
```diff
@@ -98,26 +103,74 @@ public VaultFormat8ProviderDecorator(CloudProvider delegate, CloudPath dataDir,
+			if (exception == null) {
+				return CompletableFuture.completedFuture(result);
+			} else if (exception instanceof FileAlreadyExistsException) {
+				return CompletableFuture.completedFuture(result);
```
I don't think there is a `result` in case of an `exception`. Use `folder` instead?

overheadhunter replied:
Is this covered by tests?

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/vaultformat8/VaultFormat8ProviderDecorator.java:
```diff
@@ -98,26 +103,74 @@ public VaultFormat8ProviderDecorator(CloudProvider delegate, CloudPath dataDir,
+				return getC9rPath(node).thenCompose(delegate::delete);
+			} else {
+				return deleteCiphertextDir(getDirPathFromClearTextDir(node))
+						.thenCompose(unused -> dirIdCache.evictFolderIncludingChilds(node));
```
Can be `thenRun(() -> dirIdCache.evictIncludingDescendants(node))`, see above.

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/vaultformat8/VaultFormat8ProviderDecorator.java:
```diff
@@ -169,19 +222,25 @@ private CloudItemMetadata toCleartextMetadata(CloudItemMetadata ciphertextMetada
 	}
 
-	private CloudPath getDirPath(byte[] dirId) {
+	private CloudPath getDirPathFromClearTextDir(byte[] dirId) {
```
rename to `getDirPathWithId(...)`

SailReal committed:
Add some simplifications in VaultFormat8ProviderDecorator

SailReal commented:
Thanks for the good feedback. I considered everything as useful and changed it.
To test the handler in `createFolder(...)` if the parent of the dir-folder already exists I added a new test scenario.
